PLIN-540 Go spec provides an implementation for Redis client/cache.

mrhwick opened the pull request:
Changelog:
- Added a `/redis` directory & package to house redis-related configuration objects.
- Added a `NewStandardRedisClient` function to generate a preconfigured Redis client from https://github.com/go-redis/redis
- Added a `RedisClient` object which can be used to hold onto a global reference to a generated client interface.
- Added a `NewStandardRedisCache` function to generate a preconfigured Redis cache from https://github.com/go-redis/cache
- Added a `RedisCache` object which can be used to hold onto a global reference to a generated cache interface.

mrhwick committed:
PLIN-540 Go spec provides an implementation for Redis client/cache.

micahhausler commented on redis/client.go:
```diff
@@ -0,0 +1,32 @@
+import (
+	"os"
+
+	"fmt"
```
Can you group the stdlib imports?

micahhausler commented on redis/cache.go:
```diff
@@ -0,0 +1,27 @@
+package redis
+
+import (
+	"github.com/go-redis/cache"
```
Could we use `gopkg.in/go-redis/cache.v6` instead?

micahhausler commented on redis/client.go:
```diff
@@ -0,0 +1,32 @@
+
+	"fmt"
+
+	"github.com/go-redis/redis"
```
Could we use `gopkg.in/redis.v6` here?

micahhausler commented on redis/client.go:
```diff
@@ -0,0 +1,32 @@
+var RedisClient *redis.Client
+
+// NewStandardRedisClient generates a preconfigured redis client according to our spec.
+func NewStandardRedisClient(redisHostEnvironmentVariableName string) (client *redis.Client, err error) {
```
To keep this function flexible, could it accept a [`*redis.Options`](https://godoc.org/gopkg.in/redis.v6#Options), override the `Addr` to `$REDIS_HOST:6379`, and return the client? Otherwise this package's client instantiation _has_ to be used.

The thought behind this being: if you're using our spec's package, then the you can expect the infrastructure-side contract to be fulfilled with a `REDIS_HOST` env var.

Also, we use separate Redis databases in some of our applications, and not hardcoding it here would be nice.

micahhausler commented on redis/client.go:
```diff
@@ -0,0 +1,32 @@
+
+	redisHost, present := os.LookupEnv("REDIS_HOST")
+	if !present {
+		err = fmt.Errorf("environment variable %s must be set in order to retrieve the redis hostname", redisHostEnvironmentVariableName)
```
Have you seen [`init()`](https://golang.org/doc/effective_go.html#init) functions? That may be a better place for this check to live. If you are importing this package and using it, we can do an import-time check for `REDIS_HOST`.  You should be able to replace the `os.LookupEnv(string) (string, bool)` call with `os.Getenv(string) string`

micahhausler commented on redis/cache.go:
```diff
@@ -0,0 +1,27 @@
+	}
+	codec = &cache.Codec{
+		Redis: client,
+		Marshal: func(v interface{}) ([]byte, error) {
```
Can you just set this to:
```go
Marshal:   msgpack.Marshal,
Unmarshal: msgpack.Unmarshal, 
```
Since those functions match the signature required for this parameter, there is no need for the wrapper function

micahhausler commented on redis/cache.go:
```diff
@@ -0,0 +1,27 @@
+// RedisCache is a *cache.Codec that can be set to the same cache you've created
+var RedisCache *cache.Codec
+
+// NewStandardRedisCache generates a preconfigured redis cache according to our spec.
```
Can you add to this docstring that it uses the msgpack encoder?

micahhausler reviewed (changes requested):
This is really awesome, thanks so much for doing this! I'm really glad this is getting use and providing value. 

Any chance you feel like adding a test for this? You can use [TravisCI's redis](https://docs.travis-ci.com/user/database-setup/#Redis)

mrhwick committed:
Address PR comments, reduce complexity, life is good

mrhwick committed:
Fix mistake about handling nonexistant error return

Accept *redis.Options argument.

Remove unnecessary export vars

Remove unnecessary export vars

fix mistake

micahhausler commented on redis/cache.go:
```diff
@@ -0,0 +1,17 @@
+
+// NewStandardRedisCache generates a preconfigured redis cache according to our spec, using msgpack for serialization format.
+func NewStandardRedisCache(options *redis.Options) codec *cache.Codec {
+	client := NewStandardRedisClient(options)
```
If you reeeeealy want to squeeze this down, you could inline this in your codec :)

mrhwick committed:
Merge files

mrhwick committed:
Docs

mrhwick committed:
Other doc

mrhwick committed:
Fix some little errors

mrhwick commented:
This should be good to go, albeit untested. I can add tests today, since I think I'm still blocked on using this in production for now.

mrhwick committed:
Added some tests

mrhwick committed:
Actually add the tests

micahhausler reviewed (approved):
Beautiful!
